### PR TITLE
GitHub stars data source

### PR DESCRIPTION
**Pull request by rubiojr**
Data source that imports GitHub starred repositories in JSON format.

Each starred repo is imported individually, the starred repo metadata comes from the GitHub API.

The item timestamp is set to the starred date, so they appear in the timeline the day the repo was starred.

A small json file is saved in the timeline repository data directory with the metadata retrieved from the GitHub API, which looks like:

```json
{
 "id": 841044067,
 "name": "timelinize",
 "html_url": "https://github.com/timelinize/timelinize",
 "description": "Store your data from all your accounts and devices in a single cohesive timeline on your own computer",
 "created_at": "2024-08-11T13:27:39Z",
 "updated_at": "2024-09-03T07:17:29Z",
 "pushed_at": "2024-09-02T15:31:59Z",
 "stargazers_count": 504,
 "language": "Go",
 "full_name": "timelinize/timelinize",
 "is_template": false,
 "topics": ["archival","data-archiving","data-import","timeline"],
 "private": false,
 "starred_at": "2024-08-12T17:55:48Z"
}
```

The data source currently expects the JSON file to be named like:

- `ghstars.json`
- `ghstars-<ISO date>.json`
- `ghstars-<UNIX timestamp>.json`

A new `Bookmark` class has been introduced.

### CODE REVIEW

**Comment by rubiojr** [comment]
Starred repo item render sample:

Screenshot from 2024-09-04 00-59-14

**Comment by rubiojr** [comment]
Once Timelinize is capable of using online data sources, this could fetch starred repos directly from the GitHub API.

**Review by mholt — COMMENTED** [review]
This is cool! I suppose we should add support for rendering JSON in the UI -- later... probably not a big priority for now.

Just a few comments inline below, before we merge this.

Love the bookmark classification. I think that will be useful in the future for browser bookmarks, Pocket/reader apps, etc.

And do you think JSON is the best format for the actual content of each item? I'm not opposed to that, but this'll be the first data source with JSON for its actual content. I see you're extracting many (all?) of the fields as metadata already, and I wonder if the content should simply be the URL or whatever.

But whatever the content is, we should probably keep it consistent for future items with a class of "bookmark" so that the UI can render them all consistently. Hence why I'm asking. :wink: 

This is great though, looking forward to ironing out those details and merging this!

**Comment by mholt** [comment]
One other idea I had envisioned for bookmark items is the possibility of capturing a snapshot of the page or whatever is being bookmarked and storing that as the content. That way we always have a local offline copy of it. Of course for web pages it's tricky because there's resources like scripts and images, so if we end up doing that I'm not sure how that'll work yet.

But years from now after the links are all dead the link won't mean much. Anyway, just a thought for consideration. I'd be curious to get your take on this too

File: datasources/ghstars/ghstars.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+// Data source name and ID.
+const (
+	DataSourceName = "GitHub Stars"
+	DataSourceID   = "ghstars"
+)
+
+// Options configures the data source.
+type Options struct {
+}
+
```
[the remark this answers is not in the document]
Comment (rubiojr) [reply datasources/ghstars/ghstars.go:46]: No, I wasn't aware, thanks for the tip 😃

File: datasources/ghstars/ghstars.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	"regexp"
+	"time"
+
+	"github.com/timelinize/timelinize/timeline"
+	"go.uber.org/zap"
+)
+
+// Data source name and ID.
+const (
+	DataSourceName = "GitHub Stars"
```
[the remark this answers is not in the document]
Comment (rubiojr) [reply datasources/ghstars/ghstars.go:39]: I think that's a pretty good idea. I'll change it.

File: datasources/ghstars/ghstars.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil {
+		timeline.Log.Fatal("registering data source", zap.Error(err))
+	}
+}
+
+// GHStars interacts with the file system to get items.
+type GHStars struct{}
+
+// Recognize returns whether the input file is recognized.
+func (GHStars) Recognize(_ context.Context, filenames []string) (timeline.Recognition, error) {
```
[the remark this answers is not in the document]
Comment (rubiojr) [reply datasources/ghstars/ghstars.go:81]: 👍🏾 happy to adapt the DS later if this gets merged before that refactoring.

**Comment by rubiojr** [comment]
> And do you think JSON is the best format for the actual content of each item? I'm not opposed to that, but this'll be the first data source with JSON for its actual content. I see you're extracting many (all?) of the fields as metadata already, and I wonder if the content should simply be the URL or whatever.

Yep, I wasn't sure about what to save tbh. I think your idea of saving just the URL (bookmark) is the right call, specially if this gets saved to the database and we avoid small files.

> 
> But whatever the content is, we should probably keep it consistent for future items with a class of "bookmark" so that the UI can render them all consistently. Hence why I'm asking. :wink: 

💯 agreed.

> This is great though, looking forward to ironing out those details and merging this!

🎉

**Comment by rubiojr** [comment]
> But years from now after the links are all dead the link won't mean much. Anyway, just a thought for consideration. I'd be curious to get your take on this too

I think this would be pretty awesome. Timelinize could have a background worker that iterates all the bookmarks saved every day and stores a snapshot of the page, sends it to the wayback machine API etc.

File: datasources/ghstars/ghstars.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rubiojr) [reply datasources/ghstars/ghstars.go]: Done, removed.

File: datasources/ghstars/ghstars.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rubiojr) [reply datasources/ghstars/ghstars.go]: Done, renamed to `GitHub`.

**Comment by rubiojr** [comment]
> Yep, I wasn't sure about what to save tbh. I think your idea of saving just the URL (bookmark) is the right call, specially if this gets saved to the database and we avoid small files.

Done in [442fe3b](https://github.com/timelinize/timelinize/pull/48/commits/442fe3b94b12cef1f4d2b699cab5aae74b344076), and setting content type to `text/plain`, with the side effect of a much better rendering:

Screenshot from 2024-09-04 14-27-02

Something I'd like to address, is to be able to search for GitHub stars specifically, and maybe to have a custom UI component. But given the classification is `Bookmark`, which is generic and also applies to GitHub repository bookmarks or any other bookmark, how would we distinguish when it's a GitHub bookmark (so we can have custom, more suitable rendering perhaps, similar to how GitHub does it)? How would we accomplish "list all the bookmarks that are GitHub starred repos" for example? It feels like items should also have labels or tags, or that multiple classifications should be possible? (i.e. a `GitHubStar` class). How do you envision this happening @mholt?

File: datasources/github/github.go [hunk 4]
```diff
@@ -87,6 +87,14 @@ func (GitHub) Recognize(_ context.Context, filenames []string) (timeline.Recogni
 
 // FileImport conducts an import of the data using this data source.
 func (c *GitHub) FileImport(ctx context.Context, filenames []string, itemChan chan<- *timeline.Graph, _ timeline.ListingOptions) error {
+	// If we are testing, close the channel to signal the end of the import, so we can
+	// count the total items sent.
+	defer func() {
+		if b, ok := ctx.Value("TLZ_TEST").(bool); ok && b {
+			close(itemChan)
+		}
+	}()
```
Comment (mholt) [inline datasources/github/github.go:96]: Hmm. Is it possible to have the tests close the channel when they're done? Maybe in [`T.Cleanup`](https://pkg.go.dev/testing#T.Cleanup)?

**Review by mholt — CHANGES_REQUESTED** [review]
Just did a quick pass on the latest commit -- I'm wondering if we can move channel close behavior back into the test somehow.

Looking good though!!

File: datasources/github/github.go (hunk 4 above)
Comment (rubiojr) [reply datasources/github/github.go:96]: Yeah tricky. That wouldn't work for this case, afaict. I tried a few variations without much success.

The problem I have with the channel writer not being responsible for closing it, is is that I'm not able to count the total number of items sent. The caller (reader/test) knows how many items should be written, but a bug in the writer could send less and the test would block trying to read more, or the writer could send more and the reader would stop the count after the desired number, missing the bug.

**Comment by mholt** [comment]
> I think this would be pretty awesome. Timelinize could have a background worker that iterates all the bookmarks saved every day and stores a snapshot of the page, sends it to the wayback machine API etc.
>
> I'm 💯% in the data hoarding camp 🏕️ 😃

Yeah maybe something like that could work -- kind of like how thumbnails are generated in the background after an import, maybe we can do that for bookmarks too. Store the URL during import (non-blocking), and do the download right after in the background (so as not to block imports).

I'm refactoring thumbnail tasks though so let's wait on this until I'm done with that.

I still wonder, what exactly should we download? Just the HTML page? We'd miss out on all the images and styles and scripts. What if scripts dynamically load the content? It gets messy pretty quick. We can only really store a single file for each item, but I guess that file could be an archive (zip/tar)... or maybe we could mod the processor to allow a _folder_ (a file system directory, not a single archive file) as an item's content. Or maybe that's all just too ambitious and we should stick to the root HTML file (if a web bookmark).

Any thoughts? What would the contents be for a GitHub star?

> Something I'd like to address, is to be able to search for GitHub stars specifically, and maybe to have a custom UI component. But given the classification is Bookmark, which is generic and also applies to GitHub repository bookmarks or any other bookmark, how would we distinguish when it's a GitHub bookmark (so we can have custom, more suitable rendering perhaps, similar to how GitHub does it)? How would we accomplish "list all the bookmarks that are GitHub starred repos" for example?

The search could set parameters of:

- Classification of `bookmark`
- Data source of `github`

I think this should do what you want already, unless GitHub has multiple kinds of "bookmarks" other than stars...

**Comment by mholt** [comment]
PS. I get the icons from https://tabler.io/icons -- if you want I can fill out the icon for the new classification later, or you can do it with this PR, either way. :+1: (I'd do a search for "bookmark" and just use the blue button to copy the SVG)

**Comment by rubiojr** [comment]
> I still wonder, what exactly should we download? Just the HTML page? We'd miss out on all the images and styles and scripts. What if scripts dynamically load the content? It gets messy pretty quick. We can only really store a single file for each item, but I guess that file could be an archive (zip/tar)... or maybe we could mod the processor to allow a folder (a file system directory, not a single archive file) as an item's content. Or maybe that's all just too ambitious and we should stick to the root HTML file (if a web bookmark).

> Any thoughts? What would the contents be for a GitHub star?

Yup, tricky, I've dabbled with this in the past (https://github.com/rubiojr/omnivore-exporter), and we could use an external dependency like obelisk (good and embeddable but not great) or monolith (much better), or even a headless chromium.

> I think this should do what you want already, unless GitHub has multiple kinds of "bookmarks" other than stars...

Technically it could be a github.com bookmark (i.e. a link to a markdown document somewhere or an asset), but not necessarily a repo star. But maybe good enough approximation for now.

**Comment by rubiojr** [comment]
> PS. I get the icons from https://tabler.io/icons -- if you want I can fill out the icon for the new classification later, or you can do it with this PR, either way. 👍 (I'd do a search for "bookmark" and just use the blue button to copy the SVG)

Will do!

File: datasources/github/github.go (hunk 4 above)
Comment (mholt) [reply datasources/github/github.go]: Ooo. I see. This isn't so bad, but it would be nice if it didn't require modifying the data sources like this. Let me think on it. :thinking:

Comment (rubiojr) [reply datasources/github/github.go]: @mholt I forgot an important detail that solves the issue in a nicer way without the context hack: you can read from closed channels.

**Comment by rubiojr** [comment]
> Will do! I'll take care of it.

Experimented this this in [4581e6b](https://github.com/timelinize/timelinize/pull/48/commits/4581e6ba2150778a7b0ac52102a6456306cfa3f9), but not sure if that's all it takes, looks like the timeline view could use some extra love :art: 


Screenshot from 2024-09-04 21-31-00
Screenshot from 2024-09-04 21-30-49

File: datasources/github/github.go (hunk 4 above)
Comment (mholt) [reply datasources/github/github.go]: Ohh very nice! Yes, closed channels can be read which is kinda great :)

**Review by mholt — APPROVED** [review]
This is looking good now! I think the last thing is figuring out about snapshotting the bookmark along with the item. We don't need to figure that out in this PR, but it might change some of the semantics of this data source or any others that would implement the `bookmark` classification. So if you want, we can merge this, with the understanding that it will likely change.

What if the snapshot is a PDF of the page? We'd obviously need some tool to help render that, but it could then be indexed like other PDFs, and become the content of the item, with the URL being a standard field in metadata, like "URL" or something.

**Comment by mholt** [comment]
Oh, and yeah, the timeline view is expected since each item class has to be supported by the UI. I don't yet have a way of displaying bookmarks but that's easy to add later once we decide how we want them to look.

**Comment by rubiojr** [comment]
> This is looking good now! I think the last thing is figuring out about snapshotting the bookmark along with the item. We don't need to figure that out in this PR, but it might change some of the semantics of this data source or any others that would implement the bookmark classification. So if you want, we can merge this, with the understanding that it will likely change.

@mholt I'm not in a hurry to merge any of this, so I'll follow your :heart:.

Would you like me to have a look at grabbing a PDF from a web page? or maybe integrating github.com/go-shiori/obelisk? Or should we just add all the repo metadata in JSON format that comes from the GitHub API?

**Comment by mholt** [comment]
Adding the repo metadata sounds like a good idea either way!

Hmm, I kind of envision like when you go to Print in your browser and then you choose PDF.

But a self contained HTML file would be fine too, I guess! Maybe better in some ways. So I dunno. Maybe we need to weigh pros and cons and implement one, and maybe both later and make it configurable.

**Review by mholt — CHANGES_REQUESTED** [review]
As far as getting this PR merged goes, what if we do the following:

- We should plan on being able to snapshot the bookmarked content and storing it with the item; ideally as the item's content. But since that is another task, what if for now we just continue to use the URL as the content.
- However, anticipating that in a (near-?)future change, the content of the item will be the snapshot, let's _also_ store the URL in metadata, as the "URL" field. You are already doing this. :wink:
- In other words, no changes need to be made other than the extremely nit-picky suggestions I made below. They are super not-important and probably don't matter, but I feel like the UI can title-case things if needed. :slightly_smiling_face:

**Comment by rubiojr** [comment]
> I'll create a new issue to track snapshotting bookmarks.

Sounds good! Interested in your take and the follow up! Would be nice to know if each data source should take care of snapshoting (I can see other importers/data source to also need web snapshotting services) or if the plan would be to build some kind of internal service/API for data sources to use.

**Comment by rubiojr** [comment]
>  Actually, what if we just make it be an item with no content for now. That's less confusing.

Done in e70fd0670bb092c26ecb92337800acc9f9553443

**Review by mholt — APPROVED** [review]
Awesome. Let's start with this! Thank you!!

**Comment by mholt** [comment]
@rubiojr It'd be cool to someday add support for the GitHub Account Export archive: https://github.com/settings/admin

(Just a showerthought. Don't feel obligated to do this yourself but I think it could add a lot of richness to the data!)
